Additional aggregate functions

ailegion opened the pull request:
added: GroupByAggregator - Concat function
added: GroupByAggregator - ConcatDistinct function
added: GroupByAggregator - CountNulls function
added: GroupByAggregator - LogicalAnd function
added: GroupByAggregator - LogicalOr function
added: GroupByAggregator - LongestString function
added: GroupByAggregator - ShortestString function
added: GroupByAggregator - CorrectedSumOfSquares function
added: GroupByAggregator - SumOfSquares function

Jira Task: https://issues.cask.co/browse/PLUGIN-166

ailegion committed:
Additional aggregate functions (#1)

added: GroupByAggregator - Concat function
added: GroupByAggregator - ConcatDistinct function
added: GroupByAggregator - CountNulls function
added: GroupByAggregator - EarliestDate function
added: GroupByAggregator - LatestDate function
added: GroupByAggregator - LogicalAnd function
added: GroupByAggregator - LogicalOr function
added: GroupByAggregator - LongestString function
added: GroupByAggregator - ShortestString function

* WIP: GroupByAggregator - CorrectedSumOfSquares
WIP: GroupByAggregator - SumOfSquares

* fixed: bugs with some of the of functions when merging aggregates
added: tests
updated: documentation

* fixed: SumOfSquares error message
updated: CorrectedSumOfSquares to use java Math lib

ailegion commented:
@googlebot I signed it!

albertshau commented on core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java:
```diff
@@ -61,6 +61,17 @@
     put("VARIANCE", "Variance");
     put("COLLECTLIST", "CollectList");
     put("COLLECTSET", "CollectSet");
+    put("LATESTDATE", "LatestDate");
```
why are latestdate and earliestdate needed? These seems like more specific versions of max and min, do max and min not work with dates for some reason?

ailegion replied:
this is how it is specified in the requirements doc.

bdmogal replied:
Will max and min work for dates as well? If that's the case we don't need to add more functions. But please test the scenario.

ailegion replied:
@bdmogal @albertshau tested with dates and both min and max functions worked. Removed earliest and latest date from PR.

bdmogal replied:
Yup that's fine.

ailegion committed:
removed: GroupBy EarliestDate
removed: GroupBy LatestDate

CuriousVini commented on core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/SumOfSquaresTest.java:
```diff
@@ -0,0 +1,36 @@
+import org.junit.Test;
+
+/**
+ *
```
Please add javadoc

ailegion replied:
updated

CuriousVini commented on core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/Concat.java:
```diff
@@ -0,0 +1,88 @@
+  public void mergeValue(StructuredRecord record) {
+    if (record.get(fieldName) != null) {
+      String value = Objects
+          .requireNonNull(record.get(fieldName)).toString();
```
nit: this line can move to above line

CuriousVini replied:
why do we check for `requireNonNull ` when we already have if condition?

ailegion replied:
updated

CuriousVini commented on core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/Concat.java:
```diff
@@ -0,0 +1,88 @@
+    if (record.get(fieldName) != null) {
+      String value = Objects
+          .requireNonNull(record.get(fieldName)).toString();
+      if (firstString) {
```
why only first string is being merged? This will happen in each task?

ailegion replied:
@CuriousVini this condition is only for the first string in order to avoid starting the string with the leading comma ex: "test1" instead of ", test1". The first statement after the if condition is used to merge the rest of the strings ```concatString = String.format("%s, %s", concatString, value);``` - line 62

CuriousVini replied:
Will this method get executed by all the tasks in the framework? If so this will be applied to all the splits

ailegion replied:
Correct, and split are also joined the same way - with String format - in which case also this check prevents situations like ``test1 ,, test2``.

ailegion committed:
updated: PR review fixes
removed: Redundant checks from concat functions
added: Java doc for test methods
fixed: code style formatting for all files

CuriousVini commented:
Also please open the PR against develop branch

ailegion commented:
> Also please open the PR against develop branch

new PR against develop branch: https://github.com/cdapio/hydrator-plugins/pull/1186
